File: Compose-My-Pohang-App/app/src/test/java/com/example/mypohangapp/RecommendListViewModelTest.kt
package com.example.mypohangapp

import com.example.mypohangapp.data.CategoryAndRecommendRepository
import com.example.mypohangapp.ui.RecommendListViewModel
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Test

class RecommendListViewModelTest {
    private val recommendListViewModel = RecommendListViewModel()

    @Test
    fun recommendListViewModel_SelectedCategory_CheckCategoryAndRecommendsSuccess() {
        val newCategory = CategoryAndRecommendRepository.categories[0]
        recommendListViewModel.setSelectedCategory(category = newCategory)
        val currentUiState = recommendListViewModel.uiState.value

        assertEquals(newCategory, currentUiState.selectedCategory)

        val recommend = CategoryAndRecommendRepository.recommends.filter {
            it.categoryType == currentUiState.selectedCategory?.category
        }

        assertEquals(recommend, currentUiState.recommends)
    }

    @Test
    fun recommendListViewModel_SelectedCategory_CheckCategoryAndRecommendsFail() {
        val newCategory = CategoryAndRecommendRepository.categories[0]
        recommendListViewModel.setSelectedCategory(category = newCategory)
        val currentUiState = recommendListViewModel.uiState.value

        assertNotEquals(
            CategoryAndRecommendRepository.categories[1],
            currentUiState.selectedCategory
        )

        val recommend = CategoryAndRecommendRepository.recommends.filter {
            it.categoryType == CategoryAndRecommendRepository.categories[1].category
        }

        assertNotEquals(recommend, currentUiState.recommends)
    }

    @Test
    fun recommendListViewModel_SelectedCategory_CheckCategoryAndRecommendsEdge() {
        val categories = CategoryAndRecommendRepository.categories

        categories.forEach { category ->
            recommendListViewModel.setSelectedCategory(category = category)
            val currentUiState = recommendListViewModel.uiState.value
            assertEquals(category, currentUiState.selectedCategory)

            val recommend = CategoryAndRecommendRepository.recommends.filter {
                it.categoryType == category.category
            }

            assertEquals(recommend, currentUiState.recommends)
        }
    }
}

File: SwiftUI-My-Pohang-App/My Pohang AppTests/RecommendListViewModelTest.swift
//
//  My_Pohang_AppTests.swift
//  My Pohang AppTests
//
//  Created by Jaehwa Noh on 12/23/23.
//

import XCTest
@testable import My_Pohang_App

final class RecommendListViewModelTest: XCTestCase {
    
    private var recommendListViewModel: RecommendListViewModel?
    
    override func setUpWithError() throws {
        try super.setUpWithError()
        self.recommendListViewModel = RecommendListViewModel()
    }
    
    override func tearDownWithError() throws {
        try super.tearDownWithError()
        self.recommendListViewModel = nil
    }
    
    func test_RecommendListViewModel_SelectCategory_SetCategorySuccess() {
        recommendListViewModel?.setSelectedCategory(category: CategoryAndRecommendRepository().categories[0])
        XCTAssertNotNil(recommendListViewModel?.selectedCategory)
        XCTAssertEqual(recommendListViewModel?.selectedCategory, CategoryAndRecommendRepository().categories[0])
    }
    
    func test_RecommendListViewModel_SelectCategory_SetCategoryFail() {
        recommendListViewModel?.setSelectedCategory(category: CategoryAndRecommendRepository().categories[1])
        XCTAssertNotNil(recommendListViewModel?.selectedCategory)
        XCTAssertNotEqual(recommendListViewModel?.selectedCategory, CategoryAndRecommendRepository().categories[0])
    }
    
    func test_RecommendListViewModel_SelectCategory_SetCategoryEdge() {
        CategoryAndRecommendRepository().categories.forEach { category in
            recommendListViewModel?.setSelectedCategory(category: category)
            XCTAssertNotNil(recommendListViewModel?.selectedCategory)
            XCTAssertEqual(recommendListViewModel?.selectedCategory, category)
            
            let recommends = CategoryAndRecommendRepository().recommends.filter {
                $0.categoryType == category.category
            }
            
            XCTAssertEqual(recommends, recommendListViewModel?.recommends)
        }
    }
    
    
    
    
    
}
